src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+            if diff == 0 {
+                continue;
+            }
+
+            encode_update(row, timestamp, diff, &mut buf)?;
+        }
+
+        encode_progress(upper, &mut buf)?;
+
+        let batch_name = format!("batch-{}-{}", lower, upper);
+        let batch_path = trace_path.join(&batch_name);
+        let batch_tmp_path = trace_path.join(format!("{}-tmp", batch_name));
+        // Write the file first suffixed with "-tmp" and then rename to guard against
+        // partial writes.
+        fs::write(&batch_tmp_path, buf)
+            .with_context(|| format!("failed to write batch file {}", batch_tmp_path.display()))?;
+        fs::rename(&batch_tmp_path, &batch_path).with_context(|| {
+            format!(
+                "failed to rename batch file from: {} to: {}",
+                batch_tmp_path.display(),
+                batch_path.display()
+            )
+        })?;
+
+        Ok(Batch {
+            upper,
+            lower,
+            path: batch_path,
+        })
+    }
+
+    /// Reintroduce a batch based on an available file in the trace
+    /// directory.
+    fn reinit(path: PathBuf) -> Result<Self, anyhow::Error> {
+        let batch_name = path
+            .file_name()
+            .expect("batch name known to exist")
+            .to_str()
+            .expect("batch name known to be valid utf8");
+        let parts: Vec<_> = batch_name.split('-').collect();
+        assert!(parts.len() == 3);
+        Ok(Self {
+            upper: parts[2].parse()?,
+            lower: parts[1].parse()?,
+            path,
+        })
+    }
+
+    /// Read the data from a batch stored on disk into memory.
+    fn read(&self) -> Result<Vec<Message>, anyhow::Error> {
+        read_segment(&self.path)
+    }
+
+    /// Physically concatenate all batches together into a single batch
+    /// and consolidate updates up to the frontier if provided.
+    fn compact(
+        batches: &[Batch],
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut messages = vec![];
+
+        for batch in batches {
+            messages.append(&mut read_segment(&batch.path)?);
+        }
+
+        Batch::create_from_messages(messages, trace_path, compaction_frontier)
+    }
+}
+
+/// A Trace is an on-disk representation of data meant to mimic a differential Trace.
+///
+/// A Trace listens to the `wal_path` and checks for WAL segments that it can consolidate
+/// into Batches (stored in the `trace_path`). Once it exceeds a certain number of Batches
+/// it tries to physically and logically compact them into a single batch that is
+/// compacted up to the `compaction` frontier.
+#[derive(Debug)]
+pub struct Trace {
+    trace_path: PathBuf,
+    wal_path: PathBuf,
+    batches: Vec<Batch>,
+    compaction: Option<Timestamp>,
+}
+
+impl Trace {
+    fn create(id: GlobalId, trace_path: PathBuf, wal_path: PathBuf) -> Result<Self, anyhow::Error> {
```

Comment:
I'm pro-doc-comments even on non-public methods! What is `id`? Looks like maybe just part of the error message it returns? Is that an important feature of the `Trace`?

Reply:
I don't know enough about `create_dir`, but what happens if the trace directory exists? Should this only ever be called on a fresh `trace_path`, are there requirements about `wal_path` before you call this (e.g. only on wals that haven't been collected yet)?

Reply:
added a lot more docs